src/System.Transactions/tests/EnlistTest.cs

```diff
@@ -0,0 +1,1193 @@
+                irm[i].UseSingle = true;
+
+            /* Durable RM irm[2] does on SPC, so
+			 * all volatile RMs get Rollback */
+
+            using (TransactionScope scope = new TransactionScope())
+            {
+                irm[0].Value = 2;
+                irm[1].Value = 6;
+
+                scope.Complete();
+            }
+
+            TransactionAbortedException tae = Assert.Throws<TransactionAbortedException>(() => ct.Commit());
+            Assert.IsType<NotSupportedException>(tae.InnerException);
+
+            irm[0].Check(1, 0, 0, 0, 0, 0, 0, "irm [0]");
+            irm[1].Check(0, 1, 0, 1, 0, 0, 0, "irm [1]");
+
+            InvalidOperationException ioe = Assert.Throws<InvalidOperationException>(() => ct.Commit());
+            Assert.Null(ioe.InnerException);
+
+            Transaction.Current = null;
+        }
+
+        #endregion
+
+        #region Promotable Single Phase Enlistment
+        [Fact]
+        public void Vol0_Dur0_Pspe1()
+        {
+            IntResourceManager irm = new IntResourceManager(1);
+            irm.Type = ResourceManagerType.Promotable;
+            using (TransactionScope scope = new TransactionScope())
+            {
+                irm.Value = 2;
+
+                scope.Complete();
+            }
+            irm.Check(1, 0, 0, 0, 0, 1, 0, "irm");
+        }
+
+        [Fact]
+        public void Vol1_Dur0_Pspe1()
+        {
+            IntResourceManager irm0 = new IntResourceManager(1);
+            IntResourceManager irm1 = new IntResourceManager(1);
+            irm1.Type = ResourceManagerType.Promotable;
+            using (TransactionScope scope = new TransactionScope())
+            {
+                irm0.Value = 2;
+                irm1.Value = 8;
+
+                scope.Complete();
+            }
+            irm1.Check(1, 0, 0, 0, 0, 1, 0, "irm1");
+        }
+
+        [Fact]
+        public void Vol0_Dur1_Pspe1()
+        {
+            IntResourceManager irm0 = new IntResourceManager(1);
+            IntResourceManager irm1 = new IntResourceManager(1);
+            irm0.Type = ResourceManagerType.Durable;
+            irm0.UseSingle = true;
+            irm1.Type = ResourceManagerType.Promotable;
+            using (TransactionScope scope = new TransactionScope())
+            {
+                irm0.Value = 8;
+                irm1.Value = 2;
+                Assert.Equal(0, irm1.NumEnlistFailed);
+            }
+        }
+
+        [Fact]
+        public void Vol0_Dur0_Pspe2()
+        {
+            IntResourceManager irm0 = new IntResourceManager(1);
+            IntResourceManager irm1 = new IntResourceManager(1);
+            irm0.Type = ResourceManagerType.Promotable;
+            irm1.Type = ResourceManagerType.Promotable;
+            using (TransactionScope scope = new TransactionScope())
+            {
+                irm0.Value = 8;
+                irm1.Value = 2;
+                Assert.Equal(0, irm1.NumEnlistFailed);
```

Comment:
Technically this is not correct. A call to  EnlistPromotableSinglePhase is called AFTER a successful EnlistPromotableSinglePhase for a given transaction will return "false", which should probably be considered an enlistment failure. An exception is not thrown, but the second PSPE still "failed".